tls/s2n_x509_validator.c

```diff
@@ -210,6 +212,29 @@ static uint8_t s2n_verify_host_information(struct s2n_x509_validator *validator,
             size_t name_len = (size_t) ASN1_STRING_length(current_name->d.ia5);
 
             verified = conn->verify_host_fn(name, name_len, conn->data_for_verify_host);
+        } else if (current_name->type == GEN_IPADD) {
+            /* try to validate an IP address if it's in the subject alt name. */
+            const unsigned char *ip_addr = current_name->d.iPAddress->data;
+            size_t ip_addr_len = (size_t)current_name->d.iPAddress->length;
+
+            char address[INET6_ADDRSTRLEN + 1] = {0};
+            const char *address_ptr = NULL;
+            if (ip_addr_len == 4) {
+                struct sockaddr_in s = {0};
+                s.sin_family = AF_INET;
+                memcpy(&s.sin_addr.s_addr, ip_addr, ip_addr_len);
+                address_ptr = inet_ntop(AF_INET, &s.sin_addr, address, (socklen_t) sizeof(address));
+            } else if (ip_addr_len == 16) {
+                struct sockaddr_in6 s = {0};
+                s.sin6_family = AF_INET6;
+                memcpy(&s.sin6_addr.__in6_u, ip_addr, ip_addr_len);
+                address_ptr = inet_ntop(AF_INET6, &s.sin6_addr, address, (socklen_t) sizeof(address));
```

Comment:
I feel like we need a huge warning to the verify_host_fn here that there are a lot of ways to encode an IPv6 string. Can we guarantee a canonicalization across platforms? I don't think so. The string could end up having a "::" in it , or it could be a bunch of "0:0:0" - I don't think inet_ntop makes any guarantees.

Reply:
I'm not sure what to do on this one at the moment.

Reply:
updated. Thanks for the help!